File: lib/nanosleep.rb
require "nanosleep.so"

module Nanosleep
  NS_IN_S = 1000000000
  private_constant :NS_IN_S

  NS_IN_MS = 1000000
  private_constant :NS_IN_MS

  def ns amt, hold_gvl: false
    seconds = amt / NS_IN_S
    nsec = amt % NS_IN_S
    nsleep(seconds, nsec, hold_gvl)
  end
  module_function :ns

  def ms amt, hold_gvl: false
    ns(amt * NS_IN_MS, hold_gvl:)
  end
  module_function :ms

  def s amt, hold_gvl: false
    ns(amt * NS_IN_S, hold_gvl:)
  end
  module_function :s

  module_function :nsleep
end

File: ext/nanosleep/nanosleep.c
#include <ruby.h>
#include <ruby/thread.h>
#include <time.h>

static void *
do_sleep(void * _req)
{
    struct timespec * req = (struct timespec *)_req;

    nanosleep(req, NULL);

    return NULL;
}

static VALUE
nsleep(VALUE mod, VALUE sec, VALUE nsec, VALUE hold_gvl)
{
    struct timespec req;

    req.tv_sec = NUM2INT(sec);
    req.tv_nsec = NUM2INT(nsec);

    if (RTEST(hold_gvl)) {
        do_sleep(&req);
    } else {
        rb_thread_call_without_gvl(do_sleep, &req, RUBY_UBF_PROCESS, 0);
    }

    return Qnil;
}

void Init_nanosleep() {
    VALUE mNanosleep = rb_define_module("Nanosleep");

    rb_define_method(mNanosleep, "nsleep", nsleep, 3);
}
